### PR TITLE
Allow building CBMC with multiple backend SAT solvers that can be selected at runtime via a new `--sat-solver` option

### LINKED ISSUE #7467: It should be possible to select SAT solver at runtime, not just compiletime
Currently, as of 5.73, you have the choice of building CBMC to use MiniSat or CaDiCaL as a solver. Each is faster in some cases.  However, a user cannot choose on a per-proof basis which to use: they must set the right flag when compiling CBMC then are locked into that choice.  This differs from the SMT experience, where you can choose the solver as a commandline flag

Proposal: Add a flag for selecting the SAT solver, just like in SMT

### PR DESCRIPTION
**Pull request by zhassan-aws**
Allow building CBMC with multiple backend solvers that can be selected at runtime via the new `--sat-solver` option.

For example, if CBMC is built with `cmake -S . -Bbuild -Dsat_impl="minisat2;cadical"`, then one can use `cbmc --sat-solver cadical` to use CaDiCaL instead of MiniSAT.

The release build now includes both MiniSAT2 and CaDiCaL.

Fixes: #7467 

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [X] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by TGWDB** [comment]
Assigning myself to keep up to date on this. I've done a partial review, but will wait until all checks have passed.

File: CMakeLists.txt [hunk 1]
```diff
@@ -91,7 +91,7 @@ set(enable_cbmc_tests on CACHE BOOL "Whether CBMC tests should be enabled")
 
 set(sat_impl "minisat2" CACHE STRING
     "This setting controls the SAT library which is used. Valid values are
-    'minisat2', 'glucose', 'cadical', 'ipasir-cadical' or 'ipasir-custom'"
```
Comment (NlightNFotis) [inline CMakeLists.txt]: Why was this deleted?

File: .github/workflows/pull-request-checks.yaml [hunk 2]
```diff
@@ -412,7 +412,7 @@ jobs:
         run: ccache -z --max-size=500M
       - name: Build using Make
         run: |
-          make -C src minisat2-download
```
Comment (NlightNFotis) [inline .github/workflows/pull-request-checks.yaml:485]: Why add the download to *all* of the builds even to the ones not using `cadical`?

This seems to increase build times for no observable benefit (if it's not being used).

File: COMPILING.md [hunk 3]
```diff
@@ -82,14 +82,18 @@ files.
...
-   a solver can be downloaded ahead of the above `cmake` invocation as follows:
+   the solvers can be downloaded ahead of the above `cmake` invocation as
+   follows:
    ```
    mkdir -p build/minisat2-download/minisat2-download-prefix/src/
    wget http://ftp.debian.org/debian/pool/main/m/minisat2/minisat2_2.2.1.orig.tar.gz \
      -O build/minisat2-download/minisat2-download-prefix/src/minisat2_2.2.1.orig.tar.gz
+   mkdir -p build/cadical-download/cadical-download-prefix/src/
+   wget  https://github.com/arminbiere/cadical/archive/rel-1.4.1.tar.gz \
+     -O build/cadical-download/cadical-download-prefix/src/rel-1.4.1.tar.gz
```
Comment (NlightNFotis) [inline COMPILING.md]: Note to self: candidate for automation through the `justfile`? #7486

File: COMPILING.md [hunk 4]
```diff
@@ -520,7 +520,7 @@ successfully on Windows or macOS.
 This document assumes you have already been able to build CPROVER on
 your chosen architecture.
 
-#RUNNING REGRESSION AND UNIT TESTS
+# RUNNING REGRESSION AND UNIT TESTS
```
Comment (NlightNFotis) [inline COMPILING.md:535]: I love little fixes to improve the docs 👍🏻

File: src/config.inc [hunk 5]
```diff
@@ -32,10 +32,9 @@ endif
 #MINISAT2 = ../../minisat-2.2.1
 #IPASIR = ../../ipasir
 #GLUCOSE = ../../glucose-syrup
-#CADICAL = ../../cadical
 
 # select default solver to be minisat2 if no other is specified
-ifeq ($(BOOLEFORCE)$(CHAFF)$(GLUCOSE)$(IPASIR)$(LINGELING)$(MINISAT)$(MINISAT2)$(PICOSAT)$(CADICAL),)
```
Comment (NlightNFotis) [inline src/config.inc]: Why did this need to be deleted?

It seems to also be containing `minisat2` which is the default.

File: CMakeLists.txt (hunk 1 above)
Comment (zhassan-aws) [reply CMakeLists.txt]: I removed `cadical` from the list, since the PR changes the build flow so that `cadical` is always included in the build, so there is no longer a reason to build with `-Dsat_impl=cadical`.

File: .github/workflows/pull-request-checks.yaml (hunk 2 above)
Comment (zhassan-aws) [reply .github/workflows/pull-request-checks.yaml:485]: To support the new `--cadical` option, cadical now has to be always built (otherwise, the CBMC build will fail). Downloading and building it typically takes ~11 seconds, so hopefully this doesn't add too much overhead:
```bash
$ /usr/bin/time -p make -C src cadical-download
...
real 11.28
```

File: src/config.inc (hunk 5 above)
Comment (zhassan-aws) [reply src/config.inc]: Since cadical is now always included, it was necessary to remove it from the list so that we don't prevent minisat from getting built by default as well. That way, if the build is done with `-Dsat_impl=glucose` for example, then glucose will be built in place of minisat (but cadical is always built regardless of the options).

File: .github/workflows/pull-request-checks.yaml [hunk 6]
```diff
@@ -691,7 +694,7 @@ jobs:
       - name: Zero ccache stats and limit in size
         run: ccache -z --max-size=500M
       - name: Configure CMake
-        run: cmake -S . -Bbuild -G Ninja -DCMAKE_BUILD_TYPE=Release -Dsat_impl=cadical
+        run: cmake -S . -Bbuild -G Ninja -DCMAKE_BUILD_TYPE=Release
```
Comment (tautschnig) [inline .github/workflows/pull-request-checks.yaml]: Hmm, but then we need to find a way to actually run all tests using CaDiCaL. Maybe this one should actually _not_ be changed?

File: CMakeLists.txt (hunk 1 above)
Comment (tautschnig) [reply CMakeLists.txt]: See my above comment: maybe there is, and even if it's only to run tests with CaDiCaL (without having to specify `--cadical`).

**Comment by tautschnig** [comment]
We'll have to refrain from building CaDiCaL on Windows, unless we want to use the patched version from https://github.com/deiruch/cadical (which, arguably, we could - it seems it's being kept up to date; maybe we just use this version for Windows?).

**Review by martin-cs — COMMENTED** [review]
It seems that this PR does a number of linked but different things.

A. Allow (limited) runtime switching of SAT solvers.
B. Remove the option to build without CaDiCaL.
C. Change the default solver to CaDiCaL.

A. is a good thing and *long* overdue.  I would really support this, esp. if it worked for all solvers.
B. I am less sure about but I can see it is necessary for C.
C. Has some pretty far-reaching consequences and will change pretty much the entire performance profile of the tool.  I suspect it will be for the better but it feels like the kind of change that should be profiled.

Can you explain a little more why you want this change and why it is important to do B and C?  Is `CADICAL=... make` not a option?

**Comment by zhassan-aws** [comment]
Thanks for taking a look @martin-cs.

> It seems that this PR does a number of linked but different things.
> 
> A. Allow (limited) runtime switching of SAT solvers. B. Remove the option to build without CaDiCaL. C. Change the default solver to CaDiCaL.
> 

The PR does A and B, but not C. The default SAT solver remains minisat. I updated the PR description to clarify this.

> Can you explain a little more why you want this change and why it is important to do B and C?

The point of this change is to include an option to run CBMC with cadical in CBMC releases going forward. This allows people using the CBMC binaries included in the releases to switch to cadical via an option instead of having to build CBMC from source.

**Comment by martin-cs** [comment]
> > Can you explain a little more why you want this change and why it 
> > is important to do B and C?
> 
> The point of this change is to include an option to run CBMC with
> cadical in CBMC releases going forward. This allows people using the
> CBMC binaries included in the releases to switch to cadical via an
> option instead of having to build CBMC from source.

Why not just set `CADICAL = ../../cadical` before building in the part
of the CI that builds the release binaries?  That way it is in the
release but is not added as a build dependency for everyone?  We should
also possibly use it in some of the CI jobs just so that we catch if it
breaks.

**Comment by zhassan-aws** [comment]
> Why not just set `CADICAL = ../../cadical` before building in the part of the CI that builds the release binaries?

Wouldn't that make CaDiCaL the only choice for released binaries? The point of the PR was to give users the option to switch to a different solver at runtime (cadical in this PR, but other solvers can be added later) as requested in #7467.

**Comment by martin-cs** [comment]
> > Why not just set CADICAL = ../../cadical before building in the 
> > part of the CI that builds the release binaries?
> 
> Wouldn't that make CaDiCaL the only choice for released binaries? The
> point of the PR was to give users the option to switch to a different
> solver at runtime (cadical in this PR, but other solvers can be added
> later) as requested in #7467<
> https://github.com/diffblue/cbmc/issues/7467>.

Forgive me if I have confused matters, my position is:

- I think the run-time switching is good.  It would be great to have
that for all of the solvers.  Keeping minisat2 as the default is good
too.
- I think it would be good if CADICAL remains an optional part of the
build.
- Having it enabled for the release build and some of the CI jobs seems
like a good way to get the functionality you need.

Does that make sense?

**Comment by zhassan-aws** [comment]
I see. Thanks for clarifying. I'll see if I can keep cadical an optional part of the build. In that case, we can error out if `--cadical` is passed to cbmc?

**Comment by jimgrundy** [comment]
Does CaDiCaL need to be special? Could we let people build with any/multiple solvers and then run-time switch to any of the solvers they built with?

**Comment by martin-cs** [comment]
@zhassan-aws It should be possible to guard using `HAVE_CADICAL` so that you only have the flag when it has been built like that.

@jimgrundy That would be ideal!

**Comment by zhassan-aws** [comment]
> Does CaDiCaL need to be special? 

No, but CaDiCaL was explicitly requested in #7467.

> Could we let people build with any/multiple solvers and then run-time switch to any of the solvers they built with?

That would be ideal. This PR was meant as a step in that direction.

**Comment by jimgrundy** [comment]
This is a response to a request, but it's our own request, so we have some flex here. Would it be a lot of work to do the more general thing? I mean if it is, then sure, let's do it later, but if it's not much it will be easiest to do it now while you have this in your head.

**Comment by tautschnig** [comment]
> > Could we let people build with any/multiple solvers and then run-time switch to any of the solvers they built with?
> 
> That would be ideal. This PR was meant as a step in that direction.

Should the user experience/command-line interface then perhaps be `--sat-solver <name>` with a runtime error when `<name>` is not among the solvers built with, and that same error then also listing what the valid choices are? Else I'm worried that we'll struggle to present coherent documentation (at all levels: HTML manuals and tutorials, man pages, help output).

**Comment by zhassan-aws** [comment]
> - I think it would be good if CADICAL remains an optional part of the build.

> Could we let people build with any/multiple solvers and then run-time switch to any of the solvers they built with?

I updated the PR to implement those two suggestions.

I will try to figure out the failures in the Rust API workflows tomorrow.

File: src/goto-checker/solver_factory.cpp [hunk 7]
```diff
@@ -241,6 +364,13 @@ std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_default()
   return solver;
 }
 
+void solver_factoryt::emit_solver_warning(const std::string &solver)
+{
+  messaget log(message_handler);
+  log.warning() << "The specified solver, '" << solver << "', is not available"
+                << messaget::eom;
```
Comment (tautschnig) [inline src/goto-checker/solver_factory.cpp]: 1) Could you please add a test for this case?
2) Perhaps the warning could also say that the default solver will be used instead.

File: .github/workflows/pull-request-checks.yaml (hunk 6 above)
Comment (zhassan-aws) [reply .github/workflows/pull-request-checks.yaml]: Reverted this one.

File: src/goto-checker/solver_factory.cpp (hunk 7 above)
Comment (zhassan-aws) [reply src/goto-checker/solver_factory.cpp]: Done for both.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
@zhassan-aws Could you please combine the commits such as to arrive at a set of meaningful commits? (I don't think squashing all of them into one would be appropriate, there are a bunch of independently useful changes.)

File: src/Makefile [hunk 8]
```diff
@@ -177,7 +177,7 @@ cadical-download:
 	@$(TAR) xfz $(cadical_release).tar.gz
 	@rm -Rf ../cadical
 	@mv cadical-$(cadical_release) ../cadical
-	@cd ../cadical && CXX=$(CXX) ./configure -O3 -s && make -j
+	@cd ../cadical && ./configure -O3 -s CXX="$(CXX)" CXXFLAGS=-std=c++14 && make -j cadical
```
Comment (kroening) [inline src/Makefile]: We have so far not required C++14. This is a major change, and shouldn't sneak in with a PR like this.

**Review by kroening — CHANGES_REQUESTED** [review]
Please don't sneak in the requirement for a C++14 compiler!

File: src/Makefile (hunk 8 above)
Comment (zhassan-aws) [reply src/Makefile]: This was just for compiling cadical, and nothing else. IIRC, this was needed to compile cadical with clang, because without it, it fails on this [line](https://github.com/arminbiere/cadical/blob/master/src/lookahead.cpp#L426), which cannot be compiled with a pre-C++11 standard (right angle brackets required a space in between). I think it should work with `-std=c++11` too. Would this suffice?

**Review by jimgrundy — APPROVED** [review]

File: src/Makefile (hunk 8 above)
Comment (zhassan-aws) [reply src/Makefile]: I changed it to `-std=c++11`.

**Comment by zhassan-aws** [comment]
> @zhassan-aws Could you please combine the commits such as to arrive at a set of meaningful commits? (I don't think squashing all of them into one would be appropriate, there are a bunch of independently useful changes.)

I reorganized all the changes into 5 commits. I made some more changes to `.github/workflows/pull-request-checks.yaml` as part of that.

File: src/Makefile (hunk 8 above)
Comment (kroening) [reply src/Makefile]: Sorry to be a pain here: please do the Cadical build using the Makefile system. In particular, ensure that the compilation options specified in the configuration are used.

Otherwise, anybody who wants to change any of the architectural settings has to find that line and change it there as well. We would prefer to enable architectural compiler flags to be given in a central place.

File: regression/cbmc/sat-solver/test.desc [hunk 9]
```diff
@@ -0,0 +1,9 @@
+CORE
+test.c
+--sat-solver cadical
+^EXIT=10$
+^SIGNAL=0$
+\[main\.assertion\.1] line \d+ assertion y != x: FAILURE
+\[main\.assertion\.2] line \d+ assertion y <= x: SUCCESS
+^VERIFICATION FAILED$
+--
```
Comment (TGWDB) [inline regression/cbmc/sat-solver/test.desc]: Based on the default behaviour for the `lingeling` test above, this will just use another solver if something ir wrong with CaDiCaL, so we're not actually testing that CaDiCaL is used. Can we please match the output for the actual CaDiCaL string, and/or check the failure to detect/use CaDiCaL is not present?

File: regression/cbmc/sat-solver-warning/test.desc [hunk 10]
```diff
@@ -0,0 +1,9 @@
+CORE broken-z3-smt-backend broken-cprover-smt-backend 
+test.c
+--sat-solver lingeling
+^EXIT=0$
+^SIGNAL=0$
+The specified solver, 'lingeling', is not available\. The default solver will be used instead\.
+\[main\.assertion\.1] line \d+ assertion x == y: SUCCESS
+^VERIFICATION SUCCESSFUL$
+--
```
Comment (TGWDB) [inline regression/cbmc/sat-solver-warning/test.desc:9]: This seems slightly brittle - what happens if someone adds `lingeling` to the list of built in solvers like CaDiCaL?

File: COMPILING.md [hunk 11]
```diff
@@ -371,28 +371,36 @@ compilation flag:
...
 ```
+CBMC can be built with multiple solvers, which can then be selected at runtime
+using the `--sat-solver` option.
+For example, to build CBMC with MiniSAT2 and Glucose, do:
+```
+make -C src minisat2-download glucose-download
+make -C src MINISAT2=../../minisat-2.2.1 GLUCOSE=../../glucose-syrup
+```
+In that case, MiniSAT2 will be used by default. To switch to glucose at runtime,
+run CBMC with `--sat-solver glucose`.
```
Comment (TGWDB) [inline COMPILING.md]: This implies that MiniSAT2 is the default because it appears first in the variable arguments to the `make` command. I think we should be clear on what the default (or default ordering) is after this PR is merged.

File: regression/cbmc/sat-solver-warning/test.desc (hunk 10 above)
Comment (tautschnig) [reply regression/cbmc/sat-solver-warning/test.desc:9]: Well, changing this one test seems easy enough? One option would be to extend the code that does the checks to also emit a warning if a completely different solver string is passed (none of the solvers that the code base knows about). Any other suggestions on how to avoid churn here @TGWDB?

Comment (TGWDB) [reply regression/cbmc/sat-solver-warning/test.desc:9]: That seems like a good path forward.

File: regression/cbmc/sat-solver/test.desc (hunk 9 above)
Comment (zhassan-aws) [reply regression/cbmc/sat-solver/test.desc]: I added this line to the `test.desc`:
```
Solving with CaDiCaL|The specified solver, 'cadical', is not available. The default solver will be used instead.
```
Is this sufficient?

File: regression/cbmc/sat-solver-warning/test.desc (hunk 10 above)
Comment (zhassan-aws) [reply regression/cbmc/sat-solver-warning/test.desc:9]: The PR already emits an error if the solver is none of the ones it knows about: https://github.com/diffblue/cbmc/pull/7493/files#diff-d9c9d73e84f06014b049a86daaa54f9be9744f6a3d998ac4acdc9d88989bf364R332

Anything else I need to do here?

File: COMPILING.md (hunk 11 above)
Comment (zhassan-aws) [reply COMPILING.md]: The ordering is actually specified by the order in this `#if` block:

https://github.com/diffblue/cbmc/blob/develop/src/solvers/sat/satcheck.h#L62

The first one that is included in the build becomes the default solver.

This was the case even before this PR: if CBMC is built with `make -C src GLUCOSE=../../glucose-syrup LINGELING=../../lingeling ` for example, lingeling will be used because it appears first in this `#if` block.

Should we mention this order in `COMPILING.md`? If so, which subset of solvers should we mention? Currently, only minisat2, cadical, and glucose are mentioned in `COMPILING.md`.

Comment (TGWDB) [reply COMPILING.md]: Given the open source nature of CBMC I would propose that the ordering is either:

1. Listed in the documentation explicitly (but this may go out of date if the code changes); or
2. defined in the documentation according to the code with a link to the appropriate code (which requires code basic reading ability, but is resistant to code changes).

File: regression/cbmc/sat-solver-warning/test.desc (hunk 10 above)
Comment (TGWDB) [reply regression/cbmc/sat-solver-warning/test.desc:9]: I was concerned that the tested behaviour is for lingeling being missing. Can we test with a solver name that is not existent and match that failure state? (The code is there, this is just updating the test.)

File: src/Makefile (hunk 8 above)
Comment (zhassan-aws) [reply src/Makefile]: Good point. Done.

File: regression/cbmc/sat-solver-warning/test.desc (hunk 10 above)
Comment (zhassan-aws) [reply regression/cbmc/sat-solver-warning/test.desc:9]: I added a new test (`cbmc/sat-solver-error`) that checks for the case of a non-existent solver and that cbmc exits with an error.

File: COMPILING.md (hunk 11 above)
Comment (zhassan-aws) [reply COMPILING.md]: I expanded the documentation and pointed to the code that decides the default solver.

**Review by peterschrammel — COMMENTED** [review]
Have you tried to build with `-Dsat_impl="minisat2;glucose"`?

**Comment by zhassan-aws** [comment]
> Have you tried to build with `-Dsat_impl="minisat2;glucose"`?

I don't recall if I did, but I just tried it now, and it behaves as expected:
```
$ ~/git/cbmc/build/bin/cbmc fail.c | grep Solving
Solving with MiniSAT 2.2.1 with simplifier
$ ~/git/cbmc/build/bin/cbmc fail.c --sat-solver glucose | grep Solving
Solving with Glucose Syrup with simplifier
```

**Review by peterschrammel — APPROVED** [review]

**Review by TGWDB — APPROVED** [review]

File: src/goto-checker/solver_factory.cpp [hunk 12]
```diff
@@ -213,20 +214,148 @@ make_satcheck_prop(message_handlert &message_handler, const optionst &options)
   return satcheck;
 }
 
-std::unique_ptr<solver_factoryt::solvert> solver_factoryt::get_default()
```
Comment (kroening) [inline src/goto-checker/solver_factory.cpp:216]: May I suggest to retain the name of this method. We may wish to change the default solver to be something that's not a SAT solver one day.

**Review by kroening — APPROVED** [review]

File: .github/workflows/pull-request-checks.yaml [hunk 13]
```diff
@@ -118,7 +118,7 @@ jobs:
         run: ccache -z --max-size=500M
       - name: Build with make
         run: |
-          make -C src       -j2
+          make -C src       -j2 MINISAT2=../../minisat-2.2.1 CADICAL=../../cadical
```
Comment (martin-cs) [inline .github/workflows/pull-request-checks.yaml:121]: I am probably wrong about this but shouldn't the assignment of environmental variables be before the make, as in:

```
CXXFLAGS="-O0 -g" make -j4000
```

File: .github/workflows/pull-request-checks.yaml (hunk 2 above)
Comment (martin-cs) [reply .github/workflows/pull-request-checks.yaml:485]: Is this still the case?  It would be good if it could be built without.

File: scripts/cadical-1.4.1-patch [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
++[ ! -f VERSION.txt -a ! -f ../VERSION.txt ] && \
+ die "needs to be called from build sub-directory"
+ 
+ [ -f makefile ] || \
+@@ -29,7 +29,7 @@ warning "could not find 'makefile'"
+ #--------------------------------------------------------------------------#
+ # The version.
+ #
+-VERSION="`cat ../VERSION`"
++VERSION="`cat ../VERSION.txt`"
```
Comment (martin-cs) [inline scripts/cadical-1.4.1-patch:19]: Ummm... why the rename?

File: src/goto-checker/solver_factory.h [hunk 15]
```diff
@@ -126,6 +130,7 @@ void parse_solver_options(const cmdlinet &cmdline, optionst &options);
   " --smt2                       use default SMT2 solver (Z3)\n"               \
   " --bitwuzla                   use Bitwuzla\n"                               \
   " --boolector                  use Boolector\n"                              \
+  " --sat-solver solver          use specified SAT solver\n"                   \
```
Comment (martin-cs) [inline src/goto-checker/solver_factory.h:133]: This is a terrible thing to request as it is so fiddly, so feel free to ignore BUT would it be possible to have the available solvers given here, so it prints "use specified SAT solver : minisat2, CaDiCaL" etc.

**Review by martin-cs — APPROVED** [review]
Thanks for all of the reworking on this @zhassan-aws .  I appreciate that it is tedious but I think it is really a better change because of it.

File: src/goto-checker/solver_factory.cpp (hunk 12 above)
Comment (zhassan-aws) [reply src/goto-checker/solver_factory.cpp:216]: Done

File: .github/workflows/pull-request-checks.yaml (hunk 13 above)
Comment (zhassan-aws) [reply .github/workflows/pull-request-checks.yaml:121]: AFAIK, this is a standard way for passing variables from the command-line to make (e.g. see http://books.gigatux.nl/mirror/cinanutshell/0596006977/cinanut-CHP-19-SECT-5.html). In this form, the variables are not treated as environment variables, and they are parsed by the `make` command.

File: scripts/cadical-1.4.1-patch (hunk 14 above)
Comment (zhassan-aws) [reply scripts/cadical-1.4.1-patch:19]: I added a comment [here](https://github.com/diffblue/cbmc/pull/7493/files#diff-3e2513390df543315686d7c85bd901ca9256268970032298815d2f893a9f0685R182) that explains why the rename is necessary. In short, the build failed on macos-11 without this change:

https://github.com/diffblue/cbmc/actions/runs/4138376169/jobs/7154711888#step:9:782

because the used macos-11 image is case insensitive, and so it ends up using the `VERSION` file in cadical's root directory for an `#include<version>` directive, as opposed to using the system's `version` header file. See this thread for more details:

https://github.com/marian-nmt/marian-dev/issues/871#issuecomment-891301114

File: src/goto-checker/solver_factory.h (hunk 15 above)
Comment (zhassan-aws) [reply src/goto-checker/solver_factory.h:133]: That's a good idea, but I'm not sure how easy it would be. I'll see if I can do this in a follow-up PR.

**Comment by zhassan-aws** [comment]
> Is this still the case? It would be good if it could be built without.

CaDiCaL is optional in the build, but I added it to as many workflows as possible because @tautschnig mentioned that there are a few regression tests that would be significantly sped up by using cadical, so he plans on updating them to use cadical to reduce CI time.

**Comment by thomasspriggs** [comment]
This may be something of a moot point now the PR is merged, but I don't see the Windows release package being updated in this PR to be built with multiple SAT solvers like for the Ubuntu releases. Does this mean this new functionality doesn't work on Windows or just that we have moved further further from feature parity?

**Comment by zhassan-aws** [comment]
> Does this mean this new functionality doesn't work on Windows or just that we have moved further further from feature parity?

The feature works on Windows as well, in the sense that the build can include multiple solvers. However, CaDiCaL cannot be built on Windows currently due to some platform-dependent code it uses, and thus, I haven't included it in the Windows release build. Perhaps we can try using the fork of CaDiCaL pointed out in this [comment](https://github.com/diffblue/cbmc/pull/7493#issuecomment-1400186917)? We can also try patching it ourselves.
